client-src/elements/datatypes.ts

```diff
@@ -28,3 +29,148 @@ export type LabelInfo = {
   description: string;
   variant: 'primary' | 'success' | 'neutral' | 'warning' | 'danger';
 };
+
+//form-definition
+export interface Browser {
+  chrome: {
+    bug: string;
+    blink_components: string[];
+    devrel: string[];
+    owners: string[];
+    prefixed: boolean;
+    status: {val: string};
+    desktop: string;
+    android: string;
+    webview: string;
+    ios: string;
+  };
+  ff: {
+    view: {val: string; text: string; url: string; notes: string};
+  };
+  safari: {
+    view: {val: string; text: string; url: string; notes: string};
+  };
+  webdev: {
+    view: {val: string; text: string; url: string; notes: string};
+  };
+  other: {
+    view: {notes: string};
+  };
+}
+
+export interface Resources {
+  docs: string[];
+  samples: string[];
+}
+
+export interface Standards {
+  spec: string;
+  maturity: {val: number; text: string};
+}
+
+export interface Feature {
+  category_int: number;
+  enterprise_feature_categories: string[];
+  feature_type_int: number;
+  intent_stage_int: number;
+  standards: Standards;
+  tag_review_status_int: number;
+  security_review_status_int: number;
+  privacy_review_status_int: number;
+  resources: Resources;
+  tags: string[];
+  browsers: Browser;
+}
+
+export interface FormattedFeature {
+  category: number;
+  enterprise_feature_categories: string[];
+  feature_type: number;
+  intent_stage: number;
+  accurate_as_of: boolean;
+  spec_link: string;
+  standard_maturity: number;
+  tag_review_status: number;
+  security_review_status: number;
+  privacy_review_status: number;
+  sample_links: string[];
+  doc_links: string[];
+  search_tags: string[];
+  blink_components: string;
+  bug_url: string;
+  devrel: string[];
+  owner: string[];
+  prefixed: boolean;
+  impl_status_chrome: string;
+  shipped_milestone: string;
+  shipped_android_milestone: string;
+  shipped_webview_milestone: string;
+  shipped_ios_milestone: string;
+  ff_views: string;
+  ff_views_link: string;
+  ff_views_notes: string;
+  safari_views: string;
+  safari_views_link: string;
+  safari_views_notes: string;
+  web_dev_views: string;
+  web_dev_views_link: string;
+  web_dev_views_notes: string;
+  other_views_notes: string;
+  [key: string]: any; // Allow additional properties
+}
+
+//form-field-specs
+export interface FieldAttrs {
+  title?: string;
+  type?: string;
+  multiple?: boolean;
+  placeholder?: string;
+  pattern?: string;
+  rows?: number;
+  cols?: number;
+  maxlength?: number;
+  chromedash_single_pattern?: string;
+  chromedash_split_pattern?: string;
+  disabled?: boolean;
+  min?: number;
+}
+
+export interface MilestoneRange {
+  earlier?: string;
+  later?: string;
+  allEarlier?: string;
+  allLater?: string;
+  warning?: string;
+  error?: string;
+}
+
+export interface Field {
+  type?: string;
+  name?: string;
```

Comment:
could you double check if we can use something like this:
```suggestion
  name?: keyof FormattedFeature;
```

I think we can do this. If we can, this would help tie back the possibilities of name back to FormattedFeature.

Currently, we do things like this:

```
      const featureJSONKey = ALL_FIELDS[field].name || field;
      const value = formattedFeature[featureJSONKey];
```

This reduces the benefit of typescript by doing this look up. We should narrow the type of name.


It also may need to be the union of multiple types.

```
name: keyof FormattedFeature | STAGE_FIELD_NAME_MAPPING[keyof STAGE_FIELD_NAME_MAPPING]
```

Reply:
By doing this, we could do further refactoring in the future easily by knowing what the possibilities are.

Reply:
yes, this is possible. working on it rn

Reply:
I suspect that feature names are, currently, arbitrary as long as they are unique.  There is statically defined list of feature names, and the valid feature names are only stored some data table that changes over time, which could possibly be retrieved from the server.    

So in terms of typing, I think it might be more correct to create a new type (e.g. type FeatureName = string) that happens to be an arbitrary string for now, and use that in all the places where feature names are used such as in this Field interface and in FormattedFeature, etc.  Some places it could be optional, (name: ?FeatureName) and other places it could be required (name: FeatureName).

Reply:
sounds good, just pushed new changes